### ISSUE TITLE
pcsclite: polkit policy actions not registered

### ISSUE THREAD

**Issue by foolnotion** [NixOS/nixpkgs#280826]
### Describe the bug
`pcsc-lite` uses two actions `org.debian.pcsc-lite.access_pcsc` and `org.debian.pcsc-lite.access_pcsc` do get necessary authorization via `polkit`. but these actions are not properly registed, leading to error:

```
journalctl --unit=pcscd --follow
Jan 13 23:34:19 jaghut pcscd[79235]: 00607239 auth.c:125:IsClientAuthorized() Error in authorization: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.debian.pcsc-lite.access_pcsc is not registered
Jan 13 23:34:19 jaghut pcscd[79235]: 00000005 auth.c:143:IsClientAuthorized() Process 79228 (user: 1000) is NOT authorized for action: access_pcsc
Jan 13 23:34:19 jaghut pcscd[79235]: 00000039 winscard_svc.c:355:ContextThread() Rejected unauthorized PC/SC client
Jan 13 23:34:19 jaghut pcscd[79235]: 00141489 auth.c:125:IsClientAuthorized() Error in authorization: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.debian.pcsc-lite.access_pcsc is not registered
Jan 13 23:34:19 jaghut pcscd[79235]: 00000005 auth.c:143:IsClientAuthorized() Process 79228 (user: 1000) is NOT authorized for action: access_pcsc
Jan 13 23:34:19 jaghut pcscd[79235]: 00000037 winscard_svc.c:355:ContextThread() Rejected unauthorized PC/SC client
Jan 13 23:34:19 jaghut pcscd[79235]: 00002007 auth.c:125:IsClientAuthorized() Error in authorization: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.debian.pcsc-lite.access_pcsc is not registered
```
(It can be verified that these actions don't exist by using `pkaction`)

This breaks `yubioauth-flutter` and `gpg`.

I tried both pcsclite 1.9.5 and 2.0.1 ending up with the same error.

### Steps To Reproduce
1. Set `services.pcscd.enable = true` and `security.polkit.enable = true` in `configuration.nix`, configure your smartcard (yubikey)
2. Try to use yubikey
3. Get error

### Expected behavior
There actions are registered correctly and usable.

### Notify maintainers

@anthonyroussel 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

I'm using flakes (no channels).

```console
[user@system:~]$  nix-shell -I nixpkgs=./src/nixpkgs -p nix-info --run "nix-info -m"
this path will be fetched (0.00 MiB download, 0.00 MiB unpacked):
  /nix/store/2bk3cqpj26fi3fwgshgg3k0fxwz0h55g-nix-info
copying path '/nix/store/2bk3cqpj26fi3fwgshgg3k0fxwz0h55g-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 6.7.0, NixOS, 24.05 (Uakari), 24.05.20240113.76a31fd`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `""`
 - nixpkgs: `not found`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by wrenix** [comment]
correct description of your effect, i am not able to use my yubikey with gpg.

**Comment by laxect** [comment]
This can be solved by

```nix
    environment.systemPackages = with pkgs; [
      # fix pcscd
      pcscliteWithPolkit.out
    ];

```

**Comment by foolnotion** [comment]
> This can be solved by
> 
> ```nix
>     environment.systemPackages = with pkgs; [
>       # fix pcscd
>       pcscliteWithPolkit.out
>     ];
> ```

this works, thanks! but this is really unintuitive, given that `pcscd` is already adding the package to `environment.systemPackages`: https://github.com/NixOS/nixpkgs/blob/428544ae95eec077c7f823b422afae5f174dee4b/nixos/modules/services/hardware/pcscd.nix#L55

I am leaving this open as I believe a proper fix should be provided in the module itself.

**Comment by ningw42** [comment]
Same issue here. It was working approx. last Dec. I don't have a specific commit hash. Adding `pcscliteWithPolkit.out` to `systemPackages` works for me.

```bash
Jan 15 20:41:28 yuki systemd[1]: Started PC/SC Smart Card Daemon.
Jan 15 20:41:38 yuki pcscd[485297]: 00000000 auth.c:125:IsClientAuthorized() Error in authorization: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.debian.pcsc-lite.access_pcsc is not registered
Jan 15 20:41:38 yuki pcscd[485297]: 00000016 auth.c:143:IsClientAuthorized() Process 480685 (user: 1000) is NOT authorized for action: access_pcsc
Jan 15 20:41:38 yuki pcscd[485297]: 00000052 winscard_svc.c:355:ContextThread() Rejected unauthorized PC/SC client
Jan 15 20:42:39 yuki systemd[1]: pcscd.service: Deactivated successfully.
```

```bash
❯ nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.11, NixOS, 24.05 (Uakari), 24.05.20240113.9b19f5e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels($user): `"home-manager"`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by Sorixelle** [comment]
Looks like a regression in 9ac6715549ce35ce0170a4dbf9965e477f5878b4 - `makeFlags` is implying to me that polkit actions get included in the `out` output, however:
```
nix-repl> :p pkgs.pcscliteWithPolkit.meta.outputsToInstall
[ "bin" "man" ]
```
`out` isn't included, which explains why adding `pcscliteWithPolkit.out` to `environment.systemPackages` fixes the issue.

**Comment by kamidon** [comment]
I was bitten by this when attempting to update over the weekend as well. In my case it prevented login because I use smartcards for user login which made debugging difficult due to the number of reboots required.

I tend to agree with @Sorixelle that going back to having the polkit actions file in the `bin` output instead of `out`  makes the most sense. That will mean that [this line](https://github.com/nixos/nixpkgs/blob/master/nixos/modules/services/hardware/pcscd.nix#L49) of the pcsc module will again result in the polkit action file being added to the system configuration.

**Comment by stigtsp** [comment]
> ```nix
>     environment.systemPackages = with pkgs; [
>       # fix pcscd
>       pcscliteWithPolkit.out
>     ];
> ```

This unbroke `gpg` and yubikeys for me as well

Cc: @marsam @anthonyroussel (https://github.com/NixOS/nixpkgs/pull/272529)

**Comment by foolnotion** [comment]
> Looks like a regression in [9ac6715](https://github.com/NixOS/nixpkgs/commit/9ac6715549ce35ce0170a4dbf9965e477f5878b4) (#272529) - `makeFlags` is implying to me that polkit actions get included in the `out` output, however:
> 
> ```
> nix-repl> :p pkgs.pcscliteWithPolkit.meta.outputsToInstall
> [ "bin" "man" ]
> ```
> 
> `out` isn't included, which explains why adding `pcscliteWithPolkit.out` to `environment.systemPackages` fixes the issue.
> 
> Solution ideas:
> 
>     * Add `out` to `meta.outputsToInstall`. Not sure about this one? `out` seems to also contain some libraries that probably don't need to be installed.
> 
>     * Install polkit actions to `bin` instead of `out`. IMO this one makes the most sense, especially as it seems to have been the previous behaviour before this change.

I also think the `bin` location should be fine, the `systemd` files are also there.

**Comment by 06kellyjac** [comment]
I ran into this issue when I couldn't access my yubikey anymore via pcsc for commits.

The above solution worked for me.

---

I'm going to add some comments for SEO so this is easier to find

```
gpg: selecting card failed: Service is not running
gpg: OpenPGP card not available: Service is not running


scdaemon[4644]: pcsc_establish_context failed: inserted card
```

**Comment by SuperSandro2000** [comment]
> * Add `out` to `meta.outputsToInstall`. Not sure about this one? `out` seems to also contain some libraries that probably don't need to be installed.

We want to commit this?

Otherwise I think we can close this as the fix got already merged.

**Comment by SuperSandro2000** [comment]
I've created #283452 to make the outputs a lot more expected by moving the udev file and the systemd service to the out output and the libraries to the lib output:

```
 result
├──  bin
│   └──  pcscd
├──  lib
│   └──  systemd
│       └──  system
│           ├──  pcscd.service
│           └──  pcscd.socket
├──  sbin
│   └──  pcscd
└──  share
    └──  polkit-1
        └──  actions
            └──  org.debian.pcsc-lite.policy
 result-dev
├──  bin
│   └──  pcsc-spy
├──  include
│   └──  PCSC
│       ├──  debuglog.h
│       ├──  ifdhandler.h
│       ├──  pcsclite.h
│       ├──  reader.h
│       ├──  winscard.h
│       └──  wintypes.h
├──  lib
│   └──  pkgconfig
│       └──  libpcsclite.pc
└──  nix-support
    └──  propagated-build-inputs
 result-doc
└──  share
    └──  doc
        └──  pcsc-lite
            ├──  install_spy.sh
            ├──  README.polkit
            └──  uninstall_spy.sh
 result-lib
└──  lib
    ├──  libpcsclite.la
    ├──  libpcsclite.so
    ├──  libpcsclite.so.1
    ├──  libpcsclite.so.1.0.0
    ├──  libpcscspy.la
    ├──  libpcscspy.so
    ├──  libpcscspy.so.0
    └──  libpcscspy.so.0.0.0
 result-man
└──  share
    └──  man
        ├──  man1
        │   └──  pcsc-spy.1.gz
        ├──  man5
        │   └──  reader.conf.5.gz
        └──  man8
            └──  pcscd.8.gz
```

I didn't test it yet, so it is still in draft.

If someone could verify that it doesn't break security keys again, that would be great and then please undraft it, otherwise I do it in the evening.

**Comment by SuperSandro2000** [comment]
I do not think I can fully test this due to the amount of rebuilds, so please chime in for reviews and find the one mistake I might have done (which I hope I didn't)

**Comment by foolnotion** [comment]
> I do not think I can fully test this due to the amount of rebuilds, so please chime in for reviews and find the one mistake I might have done (which I hope I didn't)

Tried to build but it fails on my machine:
```
error: builder for '/nix/store/wc590z8yj6cs043rycgz7sd5id0727kg-python3.11-dulwich-0.21.7.drv' failed with exit code 1;
       last 10 log lines:
       > dulwich/tests/test_repository.py::RepositoryRootTests::test_fetch_ignores_missing_refs
       >   /build/dulwich-0.21.7/dulwich/refs.py:1280: UserWarning: ref refs/heads/blah points at non-present sha 1234566789123456789123567891234657373833
       >     warnings.warn(
       >
       > -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
       > =========================== short test summary info ============================
       > FAILED dulwich/tests/compat/test_client.py::DulwichTCPClientTest::test_fetch_pack_no_side_band_64k - AssertionError: Lists differ: [b'07a4f52db4e20c8e63d58a342eb7f5a23b2b5d7[85...
       > FAILED dulwich/tests/compat/test_client.py::DulwichTCPClientTest::test_send_pack_nothing_to_send - AssertionError: {b'HEAD': b'514dc6d3fbfe77361bcaef320c4d21[141 chars]be9'} ...
       > ===== 2 failed, 1468 passed, 33 skipped, 1 xfailed, 251 warnings in 5.63s ======
       > /nix/store/v5irq7wvkr7kih0hhnch5nnv2dcq8c4f-stdenv-linux/setup: line 1557: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/wc590z8yj6cs043rycgz7sd5id0727kg-python3.11-dulwich-0.21.7.drv'.
error: 1 dependencies of derivation '/nix/store/mqwvy609gvix7nhc1f28nlxyx03m48w5-python3.11-breezy-3.3.4.drv' failed to build
error: 1 dependencies of derivation '/nix/store/q655714kpc1pk5g62r2vggmh5ag67wcj-nix-prefetch-bzr.drv' failed to build
error: 1 dependencies of derivation '/nix/store/9hgcfpwsf3nhi7lasbifjhsrpv9bkam0-nix-prefetch-scripts.drv' failed to build
error: 1 dependencies of derivation '/nix/store/5pgl5jvpcvp0kp6d9hss3h57r06ppnd3-system-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/0gxnj3603ybw3qirn233xa1857nmjs2x-nixos-system-jaghut-24.05.20240124.ecc103f.drv' failed to build
```

**Comment by SuperSandro2000** [comment]
I've added xdist to dulwich here https://github.com/NixOS/nixpkgs/pull/281170 . Let me see if we can just disable some tests as a speed up from 66s to 9s is quite good.

**Comment by SuperSandro2000** [comment]
This should be a lot better with #284028 . I did a lot of rebuilds locally on a low and high core machine and didn't encounter any failures anymore.

**Comment by foolnotion** [comment]
This was fixed in #283452, works fine for me.
